azurelinuxagent/ga/collect_logs.py

```diff
@@ -0,0 +1,225 @@
+
+    _THREAD_NAME = "CollectLogsHandler"
+
+    @staticmethod
+    def get_thread_name():
+        return CollectLogsHandler._THREAD_NAME
+
+    def __init__(self):
+        self.protocol = None
+        self.protocol_util = None
+        self.event_thread = None
+        self.should_run = True
+        self.last_state = None
+
+        self._periodic_operations = [
+            PeriodicOperation("collect_and_send_logs", self.collect_and_send_logs, conf.get_collect_logs_period())
+        ]
+
+    def _log_collection_allowed(self):
+        # There are three conditions that need to be met in order to allow periodic log collection:
+        # 1) It should be enabled in the configuration.
+        # 2) The system must be using systemd to manage services. Needed for resource limiting of the log collection.
+        # 3) The python version must be greater than 2.6 in order to support the ZipFile library used when collecting.
+        conf_enabled = conf.get_collect_logs()
+        systemd_present = os.path.exists("/run/systemd/system/")
+        supported_python = PY_VERSION_MINOR >= 7 if PY_VERSION_MAJOR == 2 else PY_VERSION_MAJOR == 3
+        is_allowed = conf_enabled and systemd_present and supported_python
+
+        if self.last_state != is_allowed:
+            msg = "Checking if log collection is allowed at this time [{0}]. All three conditions must be met: " \
+                  "configuration enabled [{1}], systemd present [{2}], python supported: [{3}]".format(is_allowed,
+                                                                                                       conf_enabled,
+                                                                                                       systemd_present,
+                                                                                                       supported_python)
+            self.last_state = is_allowed
+            logger.info(msg)
+            add_event(
+                name=AGENT_NAME,
+                version=CURRENT_VERSION,
+                op=WALAEventOperation.LogCollection,
+                is_success=is_allowed,
+                message=msg,
+                log_event=False)
+
+        return is_allowed
+
+    def run(self):
+        self.start(init_data=True)
+
+    def is_alive(self):
+        return self.event_thread.is_alive()
+
+    def start(self, init_data=False):
+        self.event_thread = threading.Thread(target=self.daemon, args=(init_data,))
+        self.event_thread.setDaemon(True)
+        self.event_thread.setName(self.get_thread_name())
+        self.event_thread.start()
+
+    def join(self):
+        self.event_thread.join()
+
+    def stopped(self):
+        return not self.should_run
+
+    def stop(self):
+        self.should_run = False
+        if self.is_alive():
+            self.join()
+
+    def init_protocols(self):
+        # The initialization of ProtocolUtil for the log collection thread should be done within the thread itself
```

Comment:
Why not use the same ProtocolUtil from the ExtHandler thread? Is there any internal call that can cause multithreading issues? If not then reusing the same object would be better since that is the direction we are going towards, removing all the extra ProtocolUtil objects from different places. That would also ensure not having to refresh the protocol object independently as the main loop would already do it for us.

Reply:
Although I did review the log collection thread code and I don't see any "writes" to a shared state, I think it's still better to use a separate ProtocolUtil just in case. The cost is very low, but the potential gain is high since we definitely avoid any mess with the protocol in case the log collector does modify state.

I would keep it as is since my plan when coming back is to move away from the separate thread and instead make the log collection a non-blocking call in an existing thread (yet to be decided whether monitor's or environment's thread), per Norberto's suggestion.

Reply:
>I would keep it as is since my plan when coming back is to move away from the separate thread and instead make the log collection a non-blocking call in an existing thread

Fair enough, if you plan on moving this to an existing thread anyways then you can explore this point then (make sure to capture a task for this just to ensure we dont forget about it)